src/codegen/compile.ml

```diff
@@ -3697,6 +3727,148 @@ module IC = struct
 
 end (* IC *)
 
+module StableMem = struct
+
+  let version = Int32.of_int 0
+
+  let register_globals env =
+    (* size (in pages) *)
+    E.add_global32 env "__stablemem_size" Mutable 0l
+
+  let get_mem_size env =
+    G.i (GlobalGet (nr (E.get_global env "__stablemem_size")))
+  let set_mem_size env =
+    G.i (GlobalSet (nr (E.get_global env "__stablemem_size")))
+
+  (* stable memory bounds check *)
+  let _guard env =
+    match E.mode env with
+    | Flags.ICMode | Flags.RefMode ->
+      Func.share_code1 env "__stablemem_guard"
+        ("offset", I32Type) []
+        (fun env get_offset ->
+          get_offset ^^
+          compile_unboxed_const (Int32.of_int page_size_bits) ^^
+          G.i (Binary (Wasm.Values.I32 I32Op.ShrU)) ^^
```

Comment:
See above, I believe this isn't quite correct. You'll need to add 0xffff before the shift to "round up".